packages/app-router/src/router.tsx

```diff
@@ -0,0 +1,325 @@
+ * each with the appropriate `stackId` inherited from parent routes when necessary.
+ *
+ * @remarks
+ * This function works recursively, ensuring that any nested routes are flattened into a single-level array.
+ * The `stackId` is passed down the route hierarchy, so that child routes inherit the `stackId` from their parent
+ * if they don't define their own. Routes are represented without their `children` property after flattening.
+ *
+ * @example
+ * const routes: RouteObject[] = [
+ *   {
+ *     name: 'Home',
+ *     path: '/',
+ *     Component: HomeComponent,
+ *     children: [
+ *       {
+ *         name: 'Profile',
+ *         path: '/profile',
+ *         Component: ProfileComponent,
+ *         stackId: 'mainStack',
+ *       },
+ *     ],
+ *   },
+ * ];
+ *
+ * const flattened = flatten(routes);
+ *
+ * // Output:
+ * // [
+ * //   { name: 'Home', path: '/', Component: HomeComponent },
+ * //   { name: 'Profile', path: '/profile', Component: ProfileComponent, stackId: 'mainStack' }
+ * // ]
+ */
+function flatten(
+  routes: IndexRoute[],
+  parentStackId?: string,
+): RouteChildWithoutChildren[] {
+  return routes.flatMap(route => {
+    // Extract children and the rest of the route properties
+    const {children, ...routeProps} = route as any;
+
+    // Determine the stackId to pass to child routes
+    const currentStackId =
+      (routeProps as BottomTabRoute).stackId || parentStackId;
+
+    // Create the flattened current route without children
+    const flattenedRoute = {...routeProps, stackId: currentStackId};
+
+    // Recursively flatten child routes
+    const flattenedChildren = flatten(children || [], currentStackId);
+
+    return [flattenedRoute, ...flattenedChildren];
+  });
+}
+
+/**
+ * Sets the root layout for the application and registers the app launch listener.
+ *
+ * @param {LayoutRoot} layout - The current layout object.
+ * @param {Route[]} routes - Array of route definitions.
+ * @param {Function} [onAppLaunched] - Optional callback to be called after the app is launched.
+ *
+ * @example
+ * setRootLayout(layout, routes, async () => {
+ *   console.log('App launched');
+ * });
+ */
+function setRootLayout(
+  layout: LayoutRoot,
+  routes: RouteChildWithoutChildren[],
+  onAppLaunched?: () => Promise<void>,
+): void {
+  // If no layout has been defined, set a default stack layout with the first route
+  if (!Object.keys(layout.root).length) {
+    layout.root = {
+      stack: {
+        children: [
+          {
+            component: {
+              name: routes[0]?.name!,
+            },
+          },
+        ],
+      },
+    };
+  }
+
+  // Register the app launched listener to set the root layout
+  Navigation.events().registerAppLaunchedListener(async () => {
+    try {
+      await onAppLaunched?.();
+    } catch {
+      // handle the error (e.g., log it)
+    }
+
+    Navigation.setRoot(layout);
```

Comment:
I agree with this, this would be important for conditional layouts. We should likely add an additional optional parameter called `autoRoot` or `autoLayout` - which will default to `true`. If it is `false` it will then not `setRoot` and either way it will resolve the created layout that would be used in `setRoot` based on the routes supplied.